src/runtime/hexagon_remote/qurt_thread_pool.cpp

```diff
@@ -0,0 +1,271 @@
+// After the job is done, the lock on the queue is re-acquired to update
+// the status and the thread loops again to look for new work. If a worker
+// thread doesn't find any new work, it goes to sleep until awoken by
+// the master thread.
+void worker_thread(work *owned_job) {
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+
+    bool locked = owned_job != NULL;
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    work_queue.lock();
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        work *job = NULL;
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                work_queue.unlock();
+                qurt_sem_down(&wait_for_work);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                work_queue.lock();
+                continue;
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                job = owned_job;
+            }
+        }
+        int myjob = job->next++;
+        job->active_workers++;
+        // If all tasks of the job have been claimed, then pop the job off the stack.
+        if (job->claimed()) {
+            work_queue.jobs = job->next_job;
+        }
+        // ***************************
+        // *** Work queue unlocked ***
+        // ***************************
+        work_queue.unlock();
+
+        if (!locked) {
+            int lock_status = qurt_hvx_lock(job->curr_hvx_mode);
+            if (lock_status != QURT_EOK) {
+                halide_print(job->user_context, "QuRT: qurt_hvx_lock failed (thread pool)\n");
```

Comment:
It would seem that having a fallback here to allow other threads to make progress might be safer as well. That would require a significant change as the design can't really put the work back on the queue, but perhaps waiting and then trying to lock again might work.